Added board support package for Nucleo_F446ZE

towen opened the pull request:
Includes support for USB, CAN.
Future could support: ETM Trace (if Zephyr supports it), QSPI, FMC, DCMI

Initially based on closest BSP for the same SoC: nucleo_f446re
Extra peripherals added based on closest BSP for same nucleo-144 board: nucleo_f429zi
Checked against nucleo-144 schematic
Documentation updated as best as I could

Tested: USB Device mode (samples/subsys/usb/console), Console via ST-link usart3 (samples/hello_world), user LEDs (samples/basic/blinky)
Not tested but should work: SPI, I2C, CAN
Not working yet: Quad-SPI (missing in dts/arm/st/f4/stm32f4.dtsi)
Auto-tested: ./scripts/twister -p nucleo_f446ze

My first PR - Be nice! :grin:

Signed-off-by: Tom Owen <tom.owen@zepler.net>

Commit:
board: Added board support package for Nucleo_F446ZE

Includes support for USB, CAN, ADC, DAC, and Arduino GPIO Map.
Future could support: ETM Trace (if Zephyr supports it), QSPI, FMC, DCMI

Initially based on closest BSP for the same SoC: nucleo_f446re
Extra peripherals added based on closest BSP for nucleo-144 board:
 nucleo_f429zi
Checked against nucleo-144 schematic
Documentation updated as best as I could
Arduino compatible pinmux for SPI, UART and I2C via Zio header

Tested:
  USB Device mode (samples/subsys/usb/console)
  Console via ST-link usart3 (samples/hello_world)
  user LEDs (samples/basic/blinky)
  scripts/twister --device-tests -p nucleo_f446ze
Not tested but should work: SPI, I2C, CAN, ADC, DAC
Not working yet: Quad-SPI (missing in dts/arm/st/f4/stm32f4.dtsi)

Co-authored-by: Alexandre Bourdiol
 <50730894+ABOSTM@users.noreply.github.com>
Signed-off-by: Tom Owen <tom.owen@zepler.net>

ABOSTM reviewed (changes requested):
Hi @towen,
Tank you for this Pull Request
This a a good job for a 1st one.
Few inline comments

ABOSTM commented:
@towen, 
in Zephyr, we don't create commits to fix issue found during review.
Instead you need to amend your initial commit.

By the way I figure out that one of my comment is  missing:
Arduino connector pin is different from nucleo_f446RE  see board user manual:
https://www.st.com/resource/en/data_brief/nucleo-f446ze.pdf

FRASTM commented on dts/arm/st/f4/stm32f4.dtsi:
```diff
@@ -470,6 +471,17 @@
+                        compatible = "st,stm32-qspi";
+                        #address-cells = <0x1>;
+                        #size-cells = <0x0>;
+                        reg = <0xa0001000 0x34>;
```
usually, the reg bank for the quadspi is larger ie 0x400 at 0xa0001000 even if strictly 0x34-bytes are available for the registers

towen replied:
This was copied directly from dts/arm/st/f7/stm32f7.dtsi - line last modified in 30dbc409095 
Happy to extend it to cover the whole reg bank though.

FRASTM commented on boards/arm/nucleo_f446ze/nucleo_f446ze.dts:
```diff
@@ -128,15 +128,12 @@
                      &quadspi_bk1_io0_pd11 &quadspi_bk1_io1_pd12
-                     &quadspi_bk1_io2_pf3 &quadspi_bk1_io3_pd13>;
+                     &quadspi_bk1_io2_pe2 &quadspi_bk1_io3_pd13>;
         status = "okay";
```
if not present on the nucleo_f446ze board, is it better not to enable it yet

towen commented on boards/arm/nucleo_f446ze/Kconfig.board:
```diff
@@ -0,0 +1,8 @@
+# STM32F446RE Nucleo board configuration
+
+# Copyright (c) 2018 Philémon Jaermann
```
I'm not sure if it's right to change the date of someone else's copyright notice? I have added my own notice below his to indicate that the file was added (by me) in 2021, but I left philemon's notice in place because it was present in the file I used as a template (nucleo_f446re.dts).
And indeed, `git blame` shows later modifications to that original file that did not change philemon's copyright notice. So maybe the best thing would be to assign copyright to the Zephyr project rather than individuals?

FRASTM replied:
ok

erwango replied:
Please use your own copyright here. For this kinf of files which are typically copy/paste from each other and we retain the copyright of the one making the effort to validate and upstream.

FRASTM commented:
Refer to the https://github.com/zephyrproject-rtos/zephyr/pull/37297 which is also introducing the quadspi for the stm32f4 soc family

ABOSTM commented on boards/arm/nucleo_f446ze/arduino_r3_connector.dtsi:
```diff
@@ -0,0 +1,40 @@
+
+arduino_i2c: &i2c1 {};
+arduino_spi: &spi1 {};
+arduino_serial: &usart6 {};
```
Similarly, usart6 pins should be arduino D0/D1 (RX/TX),
or this line removed. (as uart6 not enabled)

towen commented:
Thanks again ABOSTM. I have updated the DTS to map the peripherals to the Arduino pins. 
I double-checked the arduino-connector-r3.dtsi against the schematic MB1137 (nucleo-144) and it was already correct, but the DTS used different pins for the main peripherals. I have changed the pin assignments in the DTS to make use of the arduino-mapped pins.
FRASTM thanks for pointing out the other pull request. I have removed that part from my PR and referenced #37297 instead to enable QSPI

ABOSTM commented on boards/arm/nucleo_f446ze/nucleo_f446ze.dts:
```diff
@@ -0,0 +1,213 @@
+
+&spi1 {
+	pinctrl-0 = <&spi1_sck_pa5
+		     &spi1_miso_pa6 &spi1_mosi_pa7>;
```
```suggestion
		     &spi1_miso_pa6 &spi1_mosi_pa7>;
	     cs-gpios = <&gpiod 14 (GPIO_ACTIVE_LOW | GPIO_PULL_UP)>;
```

ABOSTM commented on boards/arm/nucleo_f446ze/doc/index.rst:
```diff
@@ -0,0 +1,258 @@
+- I2C1_SCL  : PB8
+- I2C2_SDA  : PF0
+- I2C2_SCL  : PF1
+- SPI1_NSS  : PA4
+- SPI1_SCK  : PB3
+- SPI1_MISO : PB4
+- SPI1_MOSI : PB5
```
Pin assignment should be changed according to changes made in DeviceTree
At least SPI1 (but could be good to check others)

ABOSTM reviewed (approved):
LGTM

erwango commented on boards/arm/nucleo_f446ze/Kconfig.defconfig:
```diff
@@ -0,0 +1,15 @@
+# STM32F446ZE Nucleo board configuration
+
+# Copyright (c) 2018 Philémon Jaermann
```
ditto

erwango commented on boards/arm/nucleo_f446ze/arduino_r3_connector.dtsi:
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2019, Christian Taedcke
```
ditto (unless pure copy/paste)

erwango commented on boards/arm/nucleo_f446ze/nucleo_f446ze.dts:
```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright (c) 2018 Philémon Jaermann
+ * Copyleft (¬c) 2021 Tom Owen
```
To my best knowledge, we don't have a copyleft policy in Zephyr, this can't be merged as is.
@cfriedt any idea on how to proceed here ?

cfriedt replied:
I've never seen this before. However, we definitely require a copy-right (name + year) as well as an SPDX license id

erwango commented on boards/arm/nucleo_f446ze/nucleo_f446ze_defconfig:
```diff
@@ -0,0 +1,27 @@
+# Enable Clocks
+CONFIG_CLOCK_CONTROL=y
+
+CONFIG_USB=y
```
Enabling is USB is application responsibility (and IIC, this was renamed recently to USB_DEVICE or something similar)

erwango reviewed (changes requested):
Thanks for this contribution!
Few comments around copyright, otherwise sounds good.

towen commented:
Sorry, I didn't mean to do that.
I had tried to retarget my PR to the latest zephyr, (so that I can un-comment the QSPI in my `nucleo_f446ze.dts` file, because QSPI has now been merged in) 
but I didn't realise it would call *everyone* 😬

cfriedt commented:
@erwango - please make a new PR - this one was merged accidentally and needs to be reverted (unless we start the v2.7.0 topic branch).
cc @nashif
